### PR TITLE
disk: maximize rootfs partition

### LINKED ISSUE #34: UCD does not resize  Clear Linux  root partition when installed on baremetal using OpenStack Ironic with a ConfigDrive
When deploying Clear Linux on baremetal using OpenStack Ironic, the root partition does not resize dynamically according to the HD size like in VM deployments. When using ConfigDrive, it is written at the end of the Hard Drive as a 64M partition, apparently this causes conflict with the resizing operation. When no ConfigDrive is used, the resize operation is executed successfully.

Steps to reproduce:

    - Build the Clear Linux cloud image with kernel-native instead of kernel-kvm https://github.com/bryteise/ister/blob/master/cloud-config.json 
    - Deploy the image with OpenStack Ironic using ConfigDrive https://docs.openstack.org/ironic/latest/install/configdrive.html

Once Clear Linux is deployed, the root partition has the same size as declared in ister, in this case 800M. 
```

root@localhost /home/clear # fdisk -l
Disk /dev/sda: 3.7 TiB, 4000225165312 bytes, 7812939776 sectors
Disk model: PERC H730P Mini 
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 1048576 bytes
Disklabel type: gpt

Device          Start        End Sectors  Size Type
/dev/sda1        2048    1048575 1046528  511M EFI System
/dev/sda2     1048576    2686975 1638400  800M Linux root (x86-64)
/dev/sda3  7812806656 7812939742  133087   65M Linux filesystem
```

### PR DESCRIPTION
**Pull request by devimc**
Other cloud-config implementations like cloud-init resizes automatically the
rootfs to the maximum allowed, ucd should do the same.

fixes #34

cc @ahkok @mdreyesm @gorozco1 

Signed-off-by: Julio Montes <julio.montes@intel.com>

### CODE REVIEW

**Comment by devimc** [comment]
@mdreyesm  please do me a favour and test it in your baremetal machine :smile:

**Comment by ahkok** [comment]
Yeah, alright, I can see this is the way to go.... Now we need testing.

File: src/disk.c [hunk 1]
```diff
@@ -63,6 +63,140 @@ static PedExceptionOption disk_exception_handler(PedException* ex) {
...
+	rootfs_dev_path = blk_device_by_path("/");
+
+	if (!rootfs_dev_path) {
+		LOG(MOD "Could not get the rootfs block device\n");
+		return NULL;
+	}
+
+	/* Looking for rootfs partition */
+	last_partition_num = ped_disk_get_last_partition_num(disk);
+	for(i = 1; i <= last_partition_num; i++) {
```
Comment (ahkok) [inline src/disk.c]:
```suggestion
	for (i = 1; i <= last_partition_num; i++) {
```

File: src/disk.c [hunk 2]
```diff
@@ -63,6 +63,140 @@ static PedExceptionOption disk_exception_handler(PedException* ex) {
...
+}
+
+static PedPartition *get_rootfs_partition(PedDisk* disk) {
+	char *rootfs_dev_path = NULL;
+	char *part_path = NULL;
+	PedPartition *part = NULL;
+	int last_partition_num = 0;
+	int i = 0;
+
+	rootfs_dev_path = blk_device_by_path("/");
```
Comment (ahkok) [inline src/disk.c]: This returns an allocated pointer, needs to be freed after use?

File: src/disk.c [hunk 3]
```diff
@@ -63,6 +63,140 @@ static PedExceptionOption disk_exception_handler(PedException* ex) {
...
+	free(rootfs_dev_path);
+	return part;
+}
+
+static gboolean resize_rootfs_blk_device(PedDisk* disk) {
+	PedPartition *rootfs_part = NULL;
+	PedGeometry geometry_start;
+	PedGeometry* geometry_end = NULL;
+	PedConstraint* constraint = NULL;
+	PedPartition* nextPartition = NULL;
```
Comment (ahkok) [inline src/disk.c]: All of these may need `free()`ing after error/exit?

File: src/disk.c (hunk 2 above)
Comment (devimc) [reply src/disk.c]: yes, here https://github.com/clearlinux/micro-config-drive/pull/35/files#diff-352f7d19b14cf77c9be0554e611634faR117

File: src/disk.c (hunk 3 above)
Comment (devimc) [reply src/disk.c]: yes, here https://github.com/clearlinux/micro-config-drive/pull/35/files#diff-352f7d19b14cf77c9be0554e611634faR192 ,  structures returned by `ped_disk_get_partition()` are not freed as we can see in this test https://fossies.org/linux/parted/libparted/tests/disk.c

File: src/disk.c (hunk 1 above)
Comment (devimc) [reply src/disk.c]: change applied, thanks

**Comment by mdreyesm** [comment]
I already tested the change deploying a Clear Linux image with the new binaries using Ironic in baremetal hosts, and it seems to work.

Thank you very much!.

```
-- Logs begin at Thu 2019-04-11 19:12:47 UTC, end at Thu 2019-04-11 19:15:10 UTC. --
Apr 11 19:12:47 clr-705d8793acc04b5e9e95117c01ba4081 systemd[1]: Starting micro-config-drive job...
Apr 11 19:12:47 clr-705d8793acc04b5e9e95117c01ba4081 ucd[479]: [7.602014] micro-config-drive version: 35
Apr 11 19:12:48 clr-705d8793acc04b5e9e95117c01ba4081 ucd[479]: [8.430918] openstack: Saving instance id '83679162-1378-4288-a2d4-70e13ec132aa'
Apr 11 19:12:48 clr-705d8793acc04b5e9e95117c01ba4081 ucd[479]: [8.431170] lib: First boot: True
Apr 11 19:12:48 clr-705d8793acc04b5e9e95117c01ba4081 ucd[479]: [8.431369] lib: Executing: /bin/sh -c "/usr/bin/usermod -p '!' root"
Apr 11 19:12:48 clr-705d8793acc04b5e9e95117c01ba4081 ucd[479]: [8.432272] Checking disk /dev/sda
Apr 11 19:12:48 clr-705d8793acc04b5e9e95117c01ba4081 ucd[479]: [8.527748] disk: Found rootfs in block device: /dev/sda2. Start sector: 1048576
Apr 11 19:12:48 clr-705d8793acc04b5e9e95117c01ba4081 ucd[479]: [8.527779] disk: Rootfs start sector: 1048576
Apr 11 19:12:48 clr-705d8793acc04b5e9e95117c01ba4081 ucd[479]: [8.527793] disk: Next partition end sector: 7812806655
Apr 11 19:12:48 clr-705d8793acc04b5e9e95117c01ba4081 ucd[479]: [8.541527] async_task: Executing [701]: /bin/sh -c "/usr/bin/resize2fs /dev/sda2"
Apr 11 19:12:48 clr-705d8793acc04b5e9e95117c01ba4081 ucd[479]: [8.541560] disk: Rootfs block device resized
```

**Comment by gorozco1** [comment]
This is great news, thanks @mdreyesm @devimc

**Comment by devimc** [comment]
@gorozco1 @mdreyesm  cool  :smile:

**Comment by ahkok** [comment]
Thanks a lot for testing this!

**Pull request merged**